beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -0,0 +1,234 @@
+	defer s.nodesLock.Unlock()
+
+	// Return if the block has been inserted into Store before.
+	if _, ok := s.nodeByRoot[root]; ok {
+		return nil
+	}
+
+	parent := s.nodeByRoot[parentRoot]
+
+	n := &Node{
+		slot:           slot,
+		root:           root,
+		parent:         parent,
+		justifiedEpoch: justifiedEpoch,
+		finalizedEpoch: finalizedEpoch,
+		optimistic:     optimistic,
+	}
+
+	s.nodeByRoot[root] = n
+	if parent != nil {
+		parent.children = append(parent.children, n)
+		if err := s.treeRootNode.updateBestDescendant(ctx, s.justifiedEpoch, s.finalizedEpoch); err != nil {
+			return err
+		}
+	}
+
+	if !optimistic {
+		if err := n.setNodeAndParentValidated(ctx); err != nil {
+			return err
+		}
+	} else {
+		optimisticCount.Inc()
+	}
+
+	// Set the node as root if the store was empty
+	if s.treeRootNode == nil {
+		s.treeRootNode = n
+		s.headNode = n
+	}
+
+	// Update metrics.
+	processedBlockCount.Inc()
+	nodeCount.Set(float64(len(s.nodeByRoot)))
+
+	return nil
+}
+
+// updateCheckpoints Update the justified / finalized epochs in store if necessary.
+func (s *Store) updateCheckpoints(justifiedEpoch, finalizedEpoch types.Epoch) {
+	s.justifiedEpoch = justifiedEpoch
+	s.finalizedEpoch = finalizedEpoch
+}
+
+// pruneMaps prunes the `nodeByRoot` map
+// starting from `node` down to the finalized Node or to a leaf of the Fork
+// choice store. This method assumes a lock on nodesLock.
+func (s *Store) pruneMaps(ctx context.Context, node, finalizedNode *Node) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+	if node == finalizedNode {
+		return nil
+	}
+	for _, child := range node.children {
+		if err := s.pruneMaps(ctx, child, finalizedNode); err != nil {
+			return err
+		}
+	}
+
+	delete(s.nodeByRoot, node.root)
+	return nil
+}
+
+// prune prunes the fork choice store with the new finalized root. The store is only pruned if the input
+// root is different than the current store finalized root, and the number of the store has met prune threshold.
+func (s *Store) prune(ctx context.Context, finalizedRoot [32]byte) error {
+	_, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.Prune")
+	defer span.End()
+
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+
+	finalizedNode, ok := s.nodeByRoot[finalizedRoot]
+	if !ok || finalizedNode == nil {
+		return errUnknownFinalizedRoot
+	}
+
+	// The number of the nodes has not met the prune threshold.
+	// Pruning at small numbers incurs more cost than benefit.
+	if finalizedNode.depth() < s.pruneThreshold {
+		return nil
+	}
+
+	// Prune nodeByRoot starting from root
+	if err := s.pruneMaps(ctx, s.treeRootNode, finalizedNode); err != nil {
+		return err
+	}
+
+	finalizedNode.parent = nil
+	s.treeRootNode = finalizedNode
+
+	prunedCount.Inc()
+	return nil
+}
+
+// tips returns a list of possible heads from fork choice store, it returns the
+// roots and the slots of the leaf nodes.
+func (s *Store) tips() ([][32]byte, []types.Slot) {
+	var roots [][32]byte
+	var slots []types.Slot
```

Comment:
We cannot know this capacity ahead of time. This depends on the number of forks. Since the happy case is just 1 and in this case the compiler does the same thing, it's better to not allocate here.